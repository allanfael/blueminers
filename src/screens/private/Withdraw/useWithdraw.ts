import { useForm } from 'react-hook-form'
import { Keyboard } from 'react-native'
import { zodResolver } from '@hookform/resolvers/zod'
import { PrivateRouteProps } from '@navigator/ParamsRoute'
import { useNavigation } from '@react-navigation/native'
import { api, WithdrawProps } from '@services/api/api'
import ResponseError from '@services/api/ResponseError'
import { Account, useAccountStore } from '@store/account'
import { roundedValue } from 'utils/currencyParse'
import { ROUTERS } from 'utils/routers'

import { withdrawFormSchema, WithdrawFormType } from './schemas'

type Props = {
  onSubmit(values: WithdrawFormType): void
  onMax(): void
  account: Account
}

export const useWithdraw = (): Props => {
  const { navigate } = useNavigation<PrivateRouteProps>()

  const { account } = useAccountStore()

  const { setError, setValue, reset } = useForm<WithdrawFormType>({
    resolver: zodResolver(withdrawFormSchema),
  })

  const onSubmit = async ({ withdraw }: WithdrawFormType) => {
    Keyboard.dismiss()

    try {
      const data: WithdrawProps = {
        account_id: account.id,
        currency: 'BRL',
        value: withdraw,
      }

      await api.withdraw(data)

      navigate(ROUTERS.WITHDRAW_CONFIRMATION, { value: Number(withdraw) })
      reset()
    } catch (e) {
      if (e instanceof ResponseError) {
        setError('withdraw', {
          message: e.message as string,
        })
      }
    }
  }

  const onMax = () => {
    if (account.available && account.available < 0) return

    setValue('withdraw', roundedValue(account.available))
  }

  return {
    onSubmit,
    onMax,
    account,
  }
}
