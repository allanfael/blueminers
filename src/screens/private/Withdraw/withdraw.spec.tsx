import React from 'react'
import { FormProvider, useForm } from 'react-hook-form'
import { render, screen } from '@testing-library/react-native'

import { useWithdraw } from './useWithdraw'
import { Withdraw } from '.'

jest.mock('./useWithdraw', () => ({
  useWithdraw: jest.fn(),
}))

jest.mock('../../../store/account', () => ({
  useAccountStore: {
    account: {
      id: '123',
      available: 100,
    },
  },
}))

describe('<Withdraw />', () => {
  it('should render Withdraw correctly', async () => {
    ;(useWithdraw as jest.MockedFunction<typeof useWithdraw>).mockReturnValue({
      onSubmit: jest.fn(),
      onMax: jest.fn(),
      account: {
        id: '123',
        available: 100,
      } as any,
    })

    render(<Withdraw />)

    expect(screen.toJSON()).toMatchSnapshot()
  })

  it('should show minimum value of withdraw', async () => {
    ;(useWithdraw as jest.MockedFunction<typeof useWithdraw>).mockReturnValue({
      onSubmit: jest.fn(),
      onMax: jest.fn(),
      account: {
        id: '123',
        available: 100,
      } as any,
    })

    const methods = useForm({
      defaultValues: { withdraw: '' },
    })

    methods.setError('withdraw', {
      message: 'Valor mínimo R$5,00',
    })

    render(
      <FormProvider {...methods}>
        <Withdraw />
      </FormProvider>,
    )

    const message = 'Valor mínimo R$5,00'

    expect(screen.getByText(message)).toBeDefined()
  })
})

// it('should render with email and password correctly', async () => {
//   ;(useLogin as jest.MockedFunction<typeof useLogin>).mockReturnValue({
//     onSubmit: jest.fn(),
//     form: {
//       formState: {
//         isSubmitting: false,
//         errors: {
//           email: {
//             message: '',
//           },
//           password: {
//             message: '',
//           },
//         },
//       },
//       handleSubmit: jest.fn(),
//     } as any,
//   })

//   const methods = useForm({
//     defaultValues: { email: '', password: '' },
//   })

//   methods.setValue('email', 'test@email')
//   methods.setValue('password', '123123')

//   render(
//     <FormProvider {...methods}>
//       <Login />
//     </FormProvider>,
//   )

//   const email = await screen.getByTestId('login-email')
//   const password = await screen.getByTestId('login-password')

//   expect(email.props.value).toBe('test@email')
//   expect(password.props.value).toBe('123123')
// })

// it('should render submit button disabled', async () => {
//   ;(useLogin as jest.MockedFunction<typeof useLogin>).mockReturnValue({
//     onSubmit: jest.fn(),
//     form: {
//       formState: {
//         isSubmitting: true,
//         errors: {
//           email: {
//             message: '',
//           },
//           password: {
//             message: '',
//           },
//         },
//       },
//       handleSubmit: jest.fn(),
//     } as any,
//   })

//   const methods = useForm({
//     defaultValues: { email: '', password: '' },
//   })

//   methods.setValue('email', 'test@email')
//   methods.setValue('password', '123123')

//   render(
//     <FormProvider {...methods}>
//       <Login />
//     </FormProvider>,
//   )

//   const button = await screen.findByTestId('login-button')

//   expect(button.props.enabled).toBeFalsy()
// })
