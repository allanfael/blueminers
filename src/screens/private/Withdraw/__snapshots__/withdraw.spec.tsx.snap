// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Withdraw /> should render Withdraw correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    [
      undefined,
      undefined,
      {
        "paddingBottom": 0,
        "paddingLeft": 30,
        "paddingRight": 30,
        "paddingTop": 20,
      },
    ]
  }
  showsVerticalScrollIndicator={false}
>
  <View>
    <Text
      style={
        {
          "color": "#000",
          "fontFamily": "Muli_700Bold",
          "fontSize": 28,
          "lineHeight": 36,
        }
      }
    >
      Saque
    </Text>
    <Text
      style={
        {
          "color": "#000",
          "fontFamily": "Muli_600SemiBold",
          "fontSize": 22,
          "lineHeight": 28,
        }
      }
    >
      Quanto você quer sacar?
    </Text>
    <View>
      <View
        style={
          [
            undefined,
            undefined,
          ]
        }
      >
        <Text
          style={
            {
              "color": "#666",
              "fontFamily": "Muli_600SemiBold",
              "fontSize": 14,
              "lineHeight": 14,
            }
          }
        >
          Mínimo: R$ 
          5,00
        </Text>
        <Text
          style={
            {
              "color": "#666",
              "fontFamily": "Muli_600SemiBold",
              "fontSize": 14,
              "lineHeight": 14,
            }
          }
        >
          Máximo: R$ 100.000,00
        </Text>
      </View>
      <View
        style={
          [
            undefined,
            {
              "backgroundColor": "#e0e0e0",
            },
          ]
        }
      >
        <TextInput
          keyboardType="numeric"
          onChangeText={[Function]}
          placeholderTextColor="#666"
          style={
            [
              undefined,
              {
                "color": "#000",
              },
            ]
          }
          testID="withdraw-input"
          value="R$0,00"
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            {
              "bottom": 12,
              "left": 12,
              "right": 12,
              "top": 12,
            }
          }
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <Text
            style={
              {
                "color": "#000",
                "fontFamily": "Muli_700Bold",
                "fontSize": 14,
                "lineHeight": 14,
              }
            }
          >
            MÁX
          </Text>
        </View>
      </View>
    </View>
    <Text
      style={
        {
          "color": "#666",
          "fontFamily": "Muli_700Bold",
          "fontSize": 14,
          "lineHeight": 14,
        }
      }
    >
      Disponível 
      R$ 100,00
    </Text>
    <View>
      <Text
        style={
          {
            "color": "#fff",
            "fontFamily": "Muli_600SemiBold",
            "fontSize": 14,
            "lineHeight": 14,
          }
        }
      >
        Ao confirmar, você irá criar sua solicitação de saque, iremos processar sua solicitação em até 2 dias úteis.
      </Text>
    </View>
    <View>
      <RNGestureHandlerButton
        activeOpacity={0.105}
        collapsable={false}
        delayLongPress={600}
        enabled={true}
        handlerTag={1}
        handlerType="NativeViewGestureHandler"
        onActiveStateChange={[Function]}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        style={
          {
            "backgroundColor": "#1f35d9",
          }
        }
        testID="submit-button"
        underlayColor="black"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "black",
              "borderBottomLeftRadius": undefined,
              "borderBottomRightRadius": undefined,
              "borderRadius": undefined,
              "borderTopLeftRadius": undefined,
              "borderTopRightRadius": undefined,
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <Text
          style={
            {
              "color": "#fff",
              "fontFamily": "Muli_700Bold",
              "fontSize": 16,
              "lineHeight": 18,
            }
          }
        >
          Confirmar
        </Text>
      </RNGestureHandlerButton>
    </View>
  </View>
</RCTScrollView>
`;
